tls/s2n_key_update.c

```diff
@@ -18,26 +18,86 @@
 #include "tls/s2n_connection.h"
 #include "tls/s2n_key_update.h"
 #include "tls/s2n_tls13_handshake.h"
+#include "tls/s2n_record.h"
 
 #include "utils/s2n_safety.h"
 
-const uint8_t S2N_KEY_UPDATE_NOT_REQUESTED = 0;
-const uint8_t S2N_KEY_UPDATE_REQUESTED = 1;
+int s2n_key_update_write(struct s2n_blob *out);
 
 int s2n_key_update_recv(struct s2n_connection *conn)
 {
     notnull_check(conn);
+
     uint8_t key_update_request;
     GUARD(s2n_stuffer_read_uint8(&conn->handshake.io, &key_update_request));
     S2N_ERROR_IF(key_update_request != S2N_KEY_UPDATE_NOT_REQUESTED && key_update_request != S2N_KEY_UPDATE_REQUESTED,
             S2N_ERR_BAD_MESSAGE);
     conn->key_update_pending = key_update_request;
+
     /* Update peer's key since a key_update was received */
     if (conn->mode == S2N_CLIENT){
         GUARD(s2n_update_application_traffic_keys(conn, S2N_SERVER, RECEIVING));
     } else {
         GUARD(s2n_update_application_traffic_keys(conn, S2N_CLIENT, RECEIVING));
     }
 
-    return 0;
+    return S2N_SUCCESS;
+}
+
+int s2n_key_update_send(struct s2n_connection *conn, size_t size) 
+{
+    notnull_check(conn);
+
+    GUARD(s2n_check_key_limits(conn, size));
+
+    if (conn->key_update_pending) {
+        uint8_t key_update_data[S2N_KEY_UPDATE_MESSAGE_SIZE];
+        struct s2n_blob key_update_blob = {0};
+        GUARD(s2n_blob_init(&key_update_blob, key_update_data, sizeof(key_update_data)));
+
+        /* Write key update message */
+        GUARD(s2n_key_update_write(&key_update_blob));
+
+        /* Encrypt the message */
+        GUARD(s2n_record_write(conn, TLS_HANDSHAKE,  &key_update_blob));
+
+        /* Update encryption key */
+        GUARD(s2n_update_application_traffic_keys(conn, conn->mode, SENDING));
+        conn->key_update_pending = false;
+        conn->encrypted_bytes_out = 0;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_key_update_write(struct s2n_blob *out)
+{
+    notnull_check(out);
+
+    struct s2n_stuffer key_update_stuffer = {0};
+    GUARD(s2n_stuffer_init(&key_update_stuffer, out));
+    GUARD(s2n_stuffer_write_uint8(&key_update_stuffer, TLS_KEY_UPDATE));
+    GUARD(s2n_stuffer_write_uint24(&key_update_stuffer, S2N_KEY_UPDATE_LENGTH));
+
+    /* s2n currently does not require peers to update their encryption keys. */
+    GUARD(s2n_stuffer_write_uint8(&key_update_stuffer, S2N_KEY_UPDATE_NOT_REQUESTED));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_check_key_limits(struct s2n_connection *conn, size_t size) 
+{
+    notnull_check(conn);
+
+    const struct s2n_record_algorithm *record_alg = conn->secure.cipher_suite->record_alg;
+    notnull_check(record_alg);
+
+    if (record_alg->encryption_limit < UINT64_MAX) {
```

Comment:
You were comparing to 0 before. Why the switch to UINT64_MAX?

Reply:
I was going to comment on this as well: because the default encryption_limit is UINT64_MAX, this isn't needed. The check inside will work without and you don't need to special case.

Reply:
Thanks y'all, I must have missed this.